plugins/microchip/switchtec-nvme.c

```diff
@@ -399,3 +400,272 @@ static int switchtec_pax_list(int argc, char **argv, struct command *command,
 
 	return 0;
 }
+
+#define NOT_FOUND	0
+#define IS_PF		1
+#define IS_VF		2
+static int pax_check_function_pdfid(struct switchtec_gfms_db_ep_port_ep *ep,
+				    uint16_t pdfid)
+{
+	int n;
+	int j;
+	int ret = NOT_FOUND;
+	struct switchtec_gfms_db_ep_port_attached_device_function *function;
+
+	n = ep->ep_hdr.function_number;
+	for (j = 0; j < n; j++) {
+		function = &ep->functions[j];
+		if (function->device_class == NVME_CLASS &&
+		    function->pdfid == pdfid) {
+			if(function->sriov_cap_pf == CAP_PF)
+				return IS_PF;
+			else if (function->sriov_cap_pf == CAP_VF)
+				return IS_VF;
+		}
+	}
+
+	return ret;
+}
+
+static int pax_check_ep_pdfid(struct switchtec_dev *dev, uint16_t pdfid)
+{
+	int i, j;
+	int ret = NOT_FOUND;
+	int n;
+	struct switchtec_gfms_db_pax_all pax_all;
+	struct switchtec_gfms_db_ep_port *ep_port;
+	struct switchtec_gfms_db_ep_port_ep *ep;
+
+	ret = switchtec_fab_gfms_db_dump_pax_all(dev, &pax_all);
+	if (ret)
+		return NOT_FOUND;
+
+	for (i = 0; i < pax_all.ep_port_all.ep_port_count; i++) {
+		ep_port = &pax_all.ep_port_all.ep_ports[i];
+		if (ep_port->port_hdr.type == SWITCHTEC_GFMS_DB_TYPE_EP) {
+			ep = &ep_port->ep_ep;
+			ret = pax_check_function_pdfid(ep, pdfid);
+			if (ret != NOT_FOUND)
+				return ret;
+		} else if (ep_port->port_hdr.type ==
+			   SWITCHTEC_GFMS_DB_TYPE_SWITCH) {
+			n = ep_port->port_hdr.ep_count;
+
+			for (j = 0; j < n; j++) {
+				ep = ep_port->ep_switch.switch_eps + j;
+				ret = pax_check_function_pdfid(ep, pdfid);
+				if (ret != NOT_FOUND)
+					return ret;
+			}
+		}
+	}
+
+	return NOT_FOUND;
+}
+
+static int pax_check_pdfid_type(struct switchtec_dev *dev, uint16_t pdfid)
+{
+	int i;
+	int ret = NOT_FOUND;
+	uint8_t r_type;
+	struct switchtec_fab_topo_info topo_info;
+	struct switchtec_gfms_db_fabric_general fg;
+
+	for(i = 0; i < SWITCHTEC_MAX_PORTS; i++)
+		topo_info.port_info_list[i].phys_port_id = 0xff;
+
+	ret = switchtec_set_pax_id(dev, SWITCHTEC_PAX_ID_LOCAL);
+	if (ret)
+		return NOT_FOUND;
+
+	ret = switchtec_topo_info_dump(dev, &topo_info);
+	if (ret)
+		return NOT_FOUND;
+
+	ret = switchtec_fab_gfms_db_dump_fabric_general(dev, &fg);
+	if (ret)
+		return NOT_FOUND;
+
+	for (i = 0; i < 16; i++) {
+		if (topo_info.route_port[i] == 0xff && fg.hdr.pax_idx != i)
+			continue;
+
+		r_type = fg.body.pax_idx[i].reachable_type;
+
+		if (fg.hdr.pax_idx == i ||
+		    r_type == SWITCHTEC_GFMS_DB_REACH_UC ||
+		    r_type == SWITCHTEC_GFMS_DB_REACH_BC) {
```

Comment:
Confirmed with firmware, we can only manage EP behind a remote PAX which connects to the local PAX with UC reachable type.